nanochat/engine.py

```diff
@@ -46,7 +46,12 @@ def eval_with_timeout(formula, max_time=3):
 def use_calculator(expr):
     """Evaluate a math expression safely."""
     expr = expr.replace(",", "")
-    if any([x not in "0123456789*+-/.() " for x in expr]): # for now disallow non-numeric chars
+    # Faster and safer check for allowed characters
+    allowed_chars = set("0123456789*+-/.() ")
+    if not all(c in allowed_chars for c in expr):
```

Comment:
I can see how this particular char-test is probably faster, but how/why is it safer?